hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/delegation/SQLDelegationTokenSecretManager.java

```diff
@@ -46,6 +50,9 @@ public abstract class SQLDelegationTokenSecretManager<TokenIdent
   private static final String SQL_DTSM_TOKEN_SEQNUM_BATCH_SIZE = SQL_DTSM_CONF_PREFIX
       + "token.seqnum.batch.size";
   public static final int DEFAULT_SEQ_NUM_BATCH_SIZE = 10;
+  public static final String SQL_DTSM_TOKEN_LOADING_CACHE_EXPIRATION_MS = SQL_DTSM_CONF_PREFIX
+      + "token.loading.cache.expiration.ms";
+  public static final int SQL_DTSM_TOKEN_LOADING_CACHE_EXPIRATION_DEFAULT_MS = 10000;
```

Comment:
Can you comment on how `SQL_DTSM_TOKEN_LOADING_CACHE_EXPIRATION_DEFAULT_MS` interacts with `DelegationTokenManager.REMOVAL_SCAN_INTERVAL_DEFAULT`. 

When the only concern is not removing renewed token, then then loading cache expiry doesn't have to be as aggressive. 10 seconds, vs. expired tokens being checked every hour.

For token cancellations though, I can see the need for a more aggressive expiry though 10 seconds may lead to too much load on the databases.

Reply:
The concern is not only with deleting tokens from SQL that have been renewed, but also with tokens being stale in the cache. If a router has a stale token that has been renewed somewhere else, the router will throw an authentication error since the token in memory is expired. We should keep this value low enough so the renewal is propagated to all routers quickly enough. Same reason for cancellations. I can see us tweaking this value to balance the impact on SQL.

The removal scan won't actually help much with token cleanup as we only expect 10 seconds of tokens to be candidates for cleanup. We need a separate mechanism to query SQL for all expired tokens and delete them, but I think we should track that separately.

Reply:
Makes sense. I hadn't thought about stale tokens.
Do you know how early renewal is done? If it is a say half-way into the token's 24 hour lifetime, then there would be more time to refresh the cache.

Anyhow, since this is a configurable value operators can explore what value is appropriate for them.

Reply:
In the tests that I've run, with a lower renew interval, I've seen renewals come 30 seconds before the expiration time. In our live service I see renewals happening 2 hours before expiration. I've yet to find how to configure it.